### PR TITLE
Add support for lists in the structured renderer

### PR DESCRIPTION
**Pull request by liamcervante**
This change is part of a series of changes that will create a new plan renderer that consumes the JSON structured output produced by the `-json` flag.

This change adds support for rendering and diffing lists.

The changes are broken up into a series of PRs that each focus on one part of the renderer to make reviewing easy. This does mean that the order matters, so you should review the PRs in the order listed below. If you prefer to view all the changes in a single (very) big PR, then skip to the last one and change it so it diffs against the main branch instead of the last PR in the chain.

The new renderer has a single entry class: `jsonformat/renderer.go`. It then makes use of two inner packages: `differ` and `change`. The `differ` package converts a single change from the provided JSON structured format into a `Change` object from the `change` package. The `change` package is then responsible for converting the `Change` objects into strings ready to be printed by the external-facing `renderer.go` functions.

This approach differs from the previous renderer where the diff calculation and the rendering happen in the same stage. Separating the diffing and rendering functionality means that going forward adding new kinds of renderers will be easier, and making changes to the format of the output plan will also be simpler.

### PR Chain

| Description         | Link                                                        | Status           |
| ------------------- | ----------------------------------------------------------- | ---------------- |
| Public JSON structs | [#32367](https://github.com/hashicorp/terraform/pull/32367) | Approved         |
| Functions and API   | [#32368](https://github.com/hashicorp/terraform/pull/32368) | Pending          |
| Primitives          | [#32373](https://github.com/hashicorp/terraform/pull/32373) | Pending          |
| Sensitive           | [#32375](https://github.com/hashicorp/terraform/pull/32375) | Pending          |
| Computed            | [#32378](https://github.com/hashicorp/terraform/pull/32378) | Pending          |
| Objects             | [#32391](https://github.com/hashicorp/terraform/pull/32391) | Pending          |
| Replace             | [#32392](https://github.com/hashicorp/terraform/pull/32392) | Pending          |
| Maps                | [#32397](https://github.com/hashicorp/terraform/pull/32397) | Pending          |
| **Lists**               | [#32401](https://github.com/hashicorp/terraform/pull/32401) | Pending          |
| Sets                | [#32409](https://github.com/hashicorp/terraform/pull/32409) | Pending          |
| Blocks              | [#32422](https://github.com/hashicorp/terraform/pull/32422) | Pending          |
| Outputs             | [#32426](https://github.com/hashicorp/terraform/pull/32426) | Pending          |

More PRs are in progress. One will port all the existing unit tests within [format/diff_test.go](https://github.com/hashicorp/terraform/blob/main/internal/command/format/diff_test.go) across to the new renderer to ensure maximum test coverage. I expect that this PR will find many missing edge cases and bugs that need to be fixed. Small bugs will be fixed as part of the main PR, but anything large will be broken out into additional PRs. Another will connect the renderer up to the rest of Terraform. This will be the last PR and it will only be merged once all possible unit tests are passing.

### CODE REVIEW

File: internal/command/jsonformat/change/list.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	var buf bytes.Buffer
+	buf.WriteString(fmt.Sprintf("[%s\n", change.forcesReplacement()))
+	for _, element := range renderer.elements {
+		if element.action == plans.NoOp && !renderNext && !opts.showUnchangedChildren {
+			unchangedElements = append(unchangedElements, element)
+			continue
+		}
+		renderNext = false
+
+		if renderer.displayContext {
```
Comment (alisdair) [inline internal/command/jsonformat/change/list.go:54]: I'm finding the logic around displaying contextual unchanged elements a bit difficult to follow, in part because it's procedural, and in part because it's Friday. Maybe some commentary on how this works would help.

It might also be neat to have a more data based approach here: take the list and the options and create a temporary set of data to be rendered, then separately render each line. This is a pie-in-the-sky idea and may not actually help, so just noting it in case it somehow inspires further thought.

File: internal/command/jsonformat/differ/list.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		if lcsIx < len(lcs) {
+			process(sliceValue.getChild(beforeIx, afterIx, false))
+			beforeIx++
+			afterIx++
+			lcsIx++
+		}
+	}
+}
+
+func lcs(xs, ys []interface{}) []interface{} {
```
Comment (alisdair) [inline internal/command/jsonformat/differ/list.go:80]: This appears to be a separate implementation of the LCS algorithm we already have in this codebase (possibly based on the other diff renderer that I wrote, or maybe just a common ancestor). We might want to refer to the one in the command/format package for now, and ideally add more commentary around its semantics and the origin of the algorithm.

**Review by alisdair — APPROVED** [review]

File: internal/command/jsonformat/change/list.go (hunk 1 above)
Comment (liamcervante) [reply internal/command/jsonformat/change/list.go:54]: I've added additional comments trying to explain things better here.

I think if I was going to redo the approach to rendering this, I'd want to make all the other collection types consistent. So I will think about this, and if I think I can improve it then I'll apply the change to all the collections in a single refactoring PR.

File: internal/command/jsonformat/differ/list.go (hunk 2 above)
Comment (liamcervante) [reply internal/command/jsonformat/differ/list.go:80]: Yes, there were two reasons I just copied across. The first was a concern over generics as the other one works over Value instead of interfaces, but I've asked about that and learned we can use generics. The second was a belief that the other one was going to be deleted when the old diff package goes away as I think that is the only place that references, so copying it over here was more of a long term move rather than a duplication. I've made a note of this, and will see if I can place it somewhere in common and get everywhere that references it to use it.
